File: rockers.py
"""
ID: totheso1
LANG: PYTHON3
TASK: rockers
"""

from typing import Any

fin = open("rockers.in", "r")
fout = open("rockers.out", "w")


def fprint(*args: Any, **kwargs: Any) -> None:
    print(*args, file=fout, **kwargs)


# Code start
SONG_COUNT, DISK_SIZE, DISK_COUNT = map(int, fin.readline().split())


def solve() -> None:
    dp = [[0] * (DISK_SIZE + 1) for _ in range(DISK_COUNT + 1)]
    for song_length in map(int, fin.readline().split()):
        for disk in reversed(range(1, DISK_COUNT + 1)):
            for curr_size in reversed(range(song_length, DISK_SIZE + 1)):
                dp[disk][curr_size] = max(
                    dp[disk][curr_size],
                    dp[disk][curr_size - song_length] + 1,
                    dp[disk - 1][DISK_SIZE] + 1,
                )
    fprint(dp[DISK_COUNT][DISK_SIZE])


solve()
# Code end

File: rockers.cpp
/*
ID: totheso1
LANG: C++14
TASK: rockers
*/

#include <algorithm>
#include <fstream>
#include <iterator>
#include <vector>

using namespace std;

static ifstream fin("rockers.in");
static ofstream fout("rockers.out");

#define endl '\n'

template <typename T>
T fin_get() {
  return *istream_iterator<T>(fin);
}

template <typename C>
C fin_get_collection(const int size) {
  C ret;
  copy_n(istream_iterator<typename C::value_type>(fin), size,
         back_inserter(ret));
  return ret;
}

static const int SONG_COUNT = fin_get<int>();
static const int DISK_SIZE = fin_get<int>();
static const int DISK_COUNT = fin_get<int>();

static void solve() {
  vector<vector<int>> dp(DISK_COUNT + 1, vector<int>(DISK_SIZE + 1, 0));
  for (int song_length; fin >> song_length;) {
    for (int disk = DISK_COUNT; disk >= 1; --disk) {
      for (int curr_size = DISK_SIZE; curr_size >= song_length; --curr_size) {
        dp[disk][curr_size] =
            max({dp[disk][curr_size], dp[disk][curr_size - song_length] + 1,
                 dp[disk - 1][DISK_SIZE] + 1});
      }
    }
  }
  fout << dp[DISK_COUNT][DISK_SIZE] << endl;
}

int main() {
  solve();
}
